### DISCUSSION TITLE
Porting cREXX to MVS/370

### DISCUSSION THREAD

**Discussion by mgrossmann** [adesutherland/CREXX#379] [Ideas]
Hello friends,

I'm thinking about whether I could get involved in the project again. As you know, I'm more of an MVS person. Could you possibly help me out where there were problems with MVS porting back then? Or could you briefly describe the build process of the VM/370 version? The topic of cREXX on MVS would actually appeal to me again.

Best regards Mike

**Comment by rvjansen** [comment]
Hi Mike,

excellent news! Are you able to join tonight's meeting?

https://us02web.zoom.us/j/85487753588?pwd=3TUyW6wmlu27hdb4mkGRVWkZIT5OSY.1￼

I think that it would be very interesting to focus on the builds for MVS and VM, in the vein of how BREXX does the cross-compiling; but let's discuss this soon with the whole team.

best regards,

René.

**Reply by mgrossmann** [reply]
Hi René, 

unfortunately  I had no chance to join the meeting. Sorry.


Best, Mike

**Comment by mgrossmann** [comment]
Hi @adesutherland 
Hi @rvjansen 

I'm not sure could help me out with my first question. :)
Today I made my first steps trying to compile for MVS.

First question, where is the nextrecLen function from?

```bash
❯ grep -ir nextrecLen   
cmsutil.c:        while ((i = nextrecLen(file)) > 0) {
```


Best, Mike

**Comment by rvjansen** [comment]
Hi Mike,I thought you need gcclib on VM. But Adrian knows.Best regards,René.On 19 Sep 2024, at 14:24, Mike Großmann ***@***.***> wrote:﻿
Hi @adesutherland
Hi @rvjansen
I'm not sure could help me out with my first question. :)
Today I made my first steps trying to compile for MVS.
First question, where is the nextrecLen function from?
❯ grep -ir nextrecLen                                                                                                                                               cmsutil.c:        while ((i = nextrecLen(file)) > 0) {                                                                                                              mvsutil.c:        while ((i = nextrecLen(file)) > 0) {
Is this a special function from the pdpclib for cms?
Best, Mike

**Reply by mgrossmann** [reply]
René, you are right. I forgot that Adrian had built a CMS version of the clib.
And yes, I found this function there. I may copy and paste this function for our mvsclib?! :) @adesutherland  

Thanks for you quick reply, René

**Reply by adesutherland** [reply]
Hello Mike!

Thank you so much for getting involved in this project again :-)

This is the comment for this function
```
/**************************************************************************************************/
/* int nextreclen(FILE * file)                                                                    */
/*                                                                                                */
/* This function returns the number of charcters the next fgets() will return                     */
/* including the /n (for text mode) but excluding the null terminator                             */
/* Non-standard GCC CMS extention - a REXX assist to a allow the input variable to be sized       */
/*                                                                                                */
/* Returns:                                                                                       */
/*    number of bytes for the next record, 0 for EOF, -1 for error                                */
/**************************************************************************************************/
```
(Noting that I can't spell!)

From https://github.com/adesutherland/CMS-370-GCCLIB/blob/8babe4674df8206e8666af3a991b8c0a166fe32a/stdio.h#L779

I think I must have added it to support brexx better

Is this being called from from code in the crexx /platform directory? We might need to make a special define for mvs to, and this is the place to put special mvs code.

My suggestion might be to start with the crexx assembler as it is relatively simple, what do you think about that?

Cheers

Adrian

**Reply by mgrossmann** [reply]
Hi @adesutherland 

I'm happy to be back, too. :) 

Yes, it was called from cmsutil from the s370 folder. 
You are right, I started with making copies of cms.h -> mvs.h and cmsutil.c -> mvsutil.c.

Let me ask more questions. :) 

Which feature branch number should I use. And from which branch should I start of? (from develop?)

Which files exactly should I generate before?  I think I need to create some parser/lexer stuff, right?

Best, Mike

**Reply by adesutherland** [reply]
I think that your work will also move forward the CMS Port.

Originally we were compiling crexx in cms, but we came to a brick wall when the compiler ran out of memory! So the plan (never executed) was to compile "locally" and then copy the mainframe assembler to the mainframe to assemble. I think this is your plan too?

I am thinking that we should make docker image to be the compiler (maybe it can build both cms and mvs, maybe not). In any case we could have docker image that creates all the mainframe assembler files. Can be used on a desktop, or on GitHub actions.

The crexx code base also includes a version of re2c (lexer) and lemon (parser) the cmake script builds these. So in fact, the dockerfile needs 

- Get the base c build environment
- Get the Regina package
- The gccmvs source tree - and build that.
- Any CMS and MVS c runtime stuff
- Do a cmake crexx build which will create the tools needed (lemon, crexx). 
- We then "just" need another cmake target for mvs and cms to compile using gccmvs to assembler.

Simples (not!) 

The included GitHub actions can help with some details.

Then you could experiment with different versions of gccmvs, gcc370, and another one on that forum based on a later version of gcc, if you like.

One bug we have is that for the threaded vm machine, the gccmvs compiler could not handle the jumps after it got passed a certain size. My theory is that the compiler is using some kind of relocatable jump with a limited range- but I don't know enough to find it (let alone getting Paul to think about helping fix it)

What do you think?

**Reply by mgrossmann** [reply]
> So the plan (never executed) was to compile "locally" and then copy the mainframe assembler to the mainframe to assemble. I think this is your plan too?

right

> I am thinking that we should make docker image to be the compiler (maybe it can build both cms and mvs, maybe not). In any case we could have docker image that creates all the mainframe assembler files. 

right, too

> Then you could experiment with different versions of gccmvs, gcc370, and another one on that forum based on a later version of gcc, if you like.

As far as I know, the 3.4.x version of gcc do have some bugs. 


So far, everything you have written sounds reasonable.

I will start trying to build the sources by hand. If this works, 
I could put my hands on the docker stuff. I'm pretty sure
that a lot of code changes will be necessary to get the MVS stuff working.

Which feature branch may I use? F0053?

Best, Mike

**Reply by adesutherland** [reply]
F0053 - Yes please
